google/cloud/storage/internal/curl_client.cc

```diff
@@ -137,6 +138,43 @@ Status CurlClient::SetupBuilder(CurlRequestBuilder& builder,
   return Status();
 }
 
+template <typename RequestType>
+std::pair<Status, std::unique_ptr<ResumableUploadSession>>
+CurlClient::CreateResumableSessionGeneric(RequestType const& request) {
+  CurlRequestBuilder builder(
+      upload_endpoint_ + "/b/" + request.bucket_name() + "/o", upload_factory_);
+  auto status = SetupBuilder(builder, request, "POST");
+  if (not status.ok()) {
+    return std::make_pair(status, std::unique_ptr<ResumableUploadSession>());
+  }
+  builder.AddQueryParameter("uploadType", "resumable");
+  builder.AddQueryParameter("name", request.object_name());
+  builder.AddHeader("Content-Type: application/json; charset=UTF-8");
+  std::string request_payload;
+  if (request.template HasOption<WithObjectMetadata>()) {
+    request_payload = request.template GetOption<WithObjectMetadata>()
+                          .value()
+                          .JsonPayloadForUpdate();
+  }
+  builder.AddHeader("Content-Length: " +
+                    std::to_string(request_payload.size()));
+  auto payload = builder.BuildRequest().MakeRequest(request_payload);
+  if (payload.status_code >= 300) {
+    return std::make_pair(
+        Status{payload.status_code, std::move(payload.payload)},
+        std::unique_ptr<ResumableUploadSession>());
+  }
+  auto response = ResumableUploadResponse::FromHttpResponse(std::move(payload));
```

Comment:
Do you think it might be more useful for debugging purposes to return the payload in the status?  I.e.

    auto response = ResumableUploadResponse::FromHttpResponse(payload);
    ....
      return std::make_pair(Status(600, std::string("Invalid server response: ") + std::move(payload)), ...);

Reply:
I liked that idea.  Unfortunately it makes the common case (when the parsing is successful) less efficient because we need to copy `payload`, which may contain the full object metadata. I implemented something slightly different. Is that good enough?

Reply:
Ah, right, we can print it out because it's preserved in the ResumableUploadResponse object that it's moved to.  Good call 👍